File: topics/tree/binary_tree.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None
class Tree:
    def __init__(self):
        self.root = None
    def inorder(self, node):
        if node is None:
            return
        self.inorder(node.left)
        print(node.data, end = " ")
        self.inorder(node.right)
    def preorder(self, node):
        if node is None:
            return
        print(node.data, end = " ")
        self.preorder(node.left)
        self.preorder(node.right)
    def postorder(self, node):
        if node is None:
            return
        self.postorder(node.left)
        self.postorder(node.right)
        print(node.data, end = " ")

tree = Tree()
tree = Tree()
tree.root = Node(1)
tree.root.left = Node(2)
tree.root.right = Node(3)
tree.root.left.left = Node(4)
tree.root.left.right = Node(5)

print("\nInorder: ")
tree.inorder(tree.root)
print("\nPreorder: ")
tree.preorder(tree.root)
print("\nPostorder: ")
tree.postorder(tree.root)

  

File: topics/tree/binary_tree.c
#include<stdio.h>
#include<stdlib.h>
typedef struct Tree{
   int data;
   struct Tree* left;
   struct Tree* right;
}Tree;
//DFS
void inorder(Tree* root){
    if(root==NULL)
    {
        return;
    }
    inorder(root->left);
    printf("%d ",root->data);
    inorder(root->right);
}
void preorder(Tree* root){
    if(root==NULL){
        return;
    }
    printf("%d ",root->data);
    preorder(root->left);
    preorder(root->right);
}
void postorder(Tree* root){
    if(root==NULL){
        return;
    }
    postorder(root->left);
    postorder(root->right);
    printf("%d ",root->data);

}

Tree* createNode(int data) {
    Tree* newNode = (Tree*)malloc(sizeof(Tree));
    newNode->data = data;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

// Example usage
int main() {
    Tree* root = createNode(1);
    root->left = createNode(2);
    root->right = createNode(3);
    root->left->left = createNode(4);
    root->left->right = createNode(5);

    printf("Inorder Traversal: ");
    inorder(root);
    printf("\n");

    printf("Preorder Traversal: ");
    preorder(root);
    printf("\n");

    printf("Postorder Traversal: ");
    postorder(root);
    printf("\n");

    return 0;
}
